File: facade-pattern/facade.cpp
// 여러개의 클래스를 하나의 클래스로 통합
#include <iostream>
using namespace std;

class CPU {
public:
    string process() {
        return "CPU 처리 중";
    }
};

class Memory {
public:
    string load() {
        return "메모리 로딩 중";
    }
};

class SSD {
public:
    string read() {
        return "SSD 드라이브 읽는 중";
    }
};

class Computer { // Facade 클래스
private:
    CPU cpu;
    Memory memory;
    SSD ssd;

public:
    void boot() {
        cout << ssd.read() << endl;
        cout << memory.load() << endl;
        cout << cpu.process() << endl;
    }
};

int main() {
    Computer c;
    c.boot();
    return 0;
}

// g++ -O3 facade.cpp -o facade
// ./facade

File: facade-pattern/facade.cs
using System;

class CPU {
    public string Process() {
        return "CPU 처리 중";
    }
}

class Memory {
    public string Load() {
        return "메모리 로딩 중";
    }
}

class SSD {
    public string Read() {
        return "SSD 드라이브 읽는 중";
    }
}

class Computer { // Facade 클래스
    private CPU cpu = new CPU();
    private Memory memory = new Memory();
    private SSD ssd = new SSD();

    public void Boot() {
        Console.WriteLine(ssd.Read());
        Console.WriteLine(memory.Load());
        Console.WriteLine(cpu.Process());
    }
}

class Program {
    static void Main(string[] args) {
        Computer c = new Computer();
        c.Boot();
    }
}

// dotnet new console -n facadeApp --force
// mv facade.cs facadeApp/Program.cs
// cd facadeApp
// dotnet run